tardis/utilities/utils.py

```diff
@@ -66,6 +67,35 @@ def csv_parser(
             }
 
 
+KeyType = TypeVar("KeyType")
+
+
+def deep_update(
+    original_mapping: MutableMapping[KeyType, Any],
+    mapping_update: Mapping[KeyType, Any],
+) -> MutableMapping[KeyType, Any]:
+    """
+    Perform a deep ``dict.update``
+
+    Recursively updates one mapping by another, merging nested ``dict``'s and
+    ``list``'s.
+    """
+
+    updated_mapping = deepcopy(original_mapping)
+    for key, value in mapping_update.items():
+        if isinstance(value, MutableMapping):
+            updated_mapping[key] = deep_update(original_mapping.get(key, {}), value)
+        elif isinstance(value, list):
+            updated_mapping[key] = deepcopy(original_mapping.get(key, []))
+            # do not duplicate entries which are already in the list
+            updated_mapping[key].extend(
```

Comment:
Is `extend` really what I expect `deep_update` to do? Running
```python
original_mapping = {"foo": [1, 2, 3], "bar": "baz"}
mapping_update = {"foo": [4, 5, 6]}
updated_mapping = deep_update(original_mapping, mapping_update)
```
I naively would expect `updated_mapping` to be `{"foo": [4, 5, 6], "bar": "baz"}` and not `{"foo": [1, 2, 3, 4, 5, 6], "bar": "baz"}`. This behavior I could achieve by explicitly running
```python
original_mapping = {"foo": [1, 2, 3], "bar": "baz"}
mapping_update = {"foo": list(set(original_mapping["foo"] + [4, 5, 6]))}
updated_mapping = deep_update(original_mapping, mapping_update)
```
Maybe this can be accomplished by explicitly setting some `extend_lists`-flag (possibly with `True` as default) in the call of `deep_update`?

Reply:
I think to implement `deep_update` was not a very wise idea and I am not happy with it as well. It seems to be prone to errors and to misunderstandings. Since, we know exactly what needs to be appended or overwritten in the Lancium SiteAdapter, I will implement it "manually" in the adapter itself following KISS.

Reply:
Have to agree that using an implementation specific to Lancium instead of trying to formulate a generic case seems more appropriate, given how many edge cases we run into otherwise. 😕